src/rgw/rgw_common.h

```diff
@@ -992,11 +992,14 @@ struct RGWBucketInfo
 
   void decode_json(JSONObj *obj);
 
-  bool versioned() { return (flags & BUCKET_VERSIONED) != 0; }
+  bool versioned() const { return (flags & BUCKET_VERSIONED) != 0; }
   int versioning_status() { return flags & (BUCKET_VERSIONED | BUCKET_VERSIONS_SUSPENDED); }
   bool versioning_enabled() { return versioning_status() == BUCKET_VERSIONED; }
 
-  bool has_swift_versioning() { return swift_versioning; }
+  bool has_swift_versioning() const {
+    /* A bucket may be versioned through one mechanism only. */
```

Comment:
I'm not sure. Although `PUT` and `COPY` doesn't seem to be a problem for the coexistence (create an archived copy first, then go to the S3 versioning), `DELETE` might be a headache. The behavior is completely different: restoring a previous version vs putting a deletion marker. This would require some arbitrage, maybe through an API extension allowing clients to pick-up a desired policy.
I'm also afraid we don't have tools for testing this increased complexity. Tempest most likely won't accept such tests. Testing some parts of Swift API in s3-tests doesn't seem to be a good option either.